src/qibo/derivative.py

```diff
@@ -26,13 +28,23 @@ def parameter_shift(
     variational parameters we need to "free" this procedure from the x depencency.
     If the `scale_factor` is not provided, it is set equal to one and doesn't
     affect the calculation.
+    If the PSR is needed to be executed on a real quantum device, it is important
+    to set `nshots` to some integer value. This enables the execution on the
+    hardware by calling the proper methods.
 
     Args:
         circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit.
         hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): target observable.
-        parameter_index (int): the index which identifies the target parameter in the circuit.get_parameters() list
+        if you want to execute on hardware, a symbolic hamiltonian must be provided
+        as follows (example with Pauli Z and `nqubits=1`): `SymbolicHamiltonian(np.prod([ Z(i) for i in range(1) ]))`.
+        parameter_index (int): the index which identifies the target parameter in the `circuit.get_parameters()` list.
         initial_state ((2**nqubits) vector): initial state on which the circuit acts (default None).
         scale_factor (float): parameter scale factor (default None).
+        nshots (int): number of shots if derivative is evaluated on hardware. If
+        `None`, the simulation mode is executed (default None).
+        nruns (int): number of times the derivative is evaluated on hardware for
+        a more stable evaluation. The final result is the mean of the values in this case
+        (default 1).
```

Comment:
This is fine, but I think if I were to deploy this on hardware using multiple runs (software averages), maybe I would do a loop over the whole method myself, eg:
```py
grads = [parameter_shift(...) for _ in range(nruns)]
```
because this way I get each individual calculation so I can calculate more statistics than just the mean (eg. `np.std(grads)` could give an estimate for the error / error bars). If I use the `nruns` I only get the mean and I don't have access to each individual sample.

That being said, having it there does not really hurt, as I can still use `nruns=1` and do the above. At the end it is the same discussion with `software_averages` argument in qibocal.